File: scripts/global/friction-event.js
'use strict';
// friction-event.js — structured per-team friction-event schema (#3165, Epic #3008/#3095).
//
// A "friction" is any harness pothole an operator hits mid-flight: a gate false-block, a
// workaround, a retry, a fleet timeout. Today these are captured ad hoc in prose (e.g. the
// Cursor Team's 18-item manual log on #3008) or lost, so cross-team recurrence goes
// undetected. emitFriction writes a REDACTED v3 event tagged `tier: 1` + `pattern_id` +
// `severity` to incidents.jsonl, so the existing anneal-tier2-autofile recurrence model
// (classifyTier1: count >= 2 by pattern_id) routes recurring frictions to Tier-2 with no
// extra plumbing. Cross-team weighting lives in friction-recurrence.js.

const os = require('os');
const path = require('path');
const fs = require('fs');
const { V3, emitV3 } = require('./event-schema-v3');
const { redactEvent } = require('./log-redaction');

const INCIDENTS_FILE = path.join(os.homedir(), '.megingjord', 'incidents.jsonl');
const CATALOG_FILE = path.join(__dirname, '..', '..', 'config', 'friction-pattern-catalog.json');
const FRICTION_EVENT = 'governance.friction';
const SEVERITIES = ['low', 'medium', 'high', 'critical'];
const DEFAULT_SEVERITY = 'low';

function buildFrictionEvent(patternId, fields = {}, now) {
  const severity = SEVERITIES.includes(fields.severity) ? fields.severity : DEFAULT_SEVERITY;
  const event = {
    version: V3,
    ts: now || new Date().toISOString(),
    service: 'friction',
    env: fields.env || 'local',
    event: FRICTION_EVENT,
    tier: 1, // tier:1 => consumed by the anneal-tier2-autofile recurrence detector
    pattern_id: patternId,
    team: fields.team || 'unknown',
    runtime: fields.runtime || 'unknown',
    role: fields.role || null,
    surface: fields.surface || null,
    severity,
    workaround: fields.workaround || null,
    trigger_role: fields.role || 'system',
    trigger_type: 'friction',
  };
  if (fields.cost !== undefined && fields.cost !== null) event.cost = fields.cost;
  if (fields.detail) event.detail = fields.detail;
  return event;
}

function isValidFriction(event) {
  const errors = [];
  if (!event || typeof event !== 'object') return { ok: false, errors: ['event must be an object'] };
  if (event.event !== FRICTION_EVENT) errors.push(`event must be ${FRICTION_EVENT}`);
  if (!event.pattern_id) errors.push('missing pattern_id');
  if (event.tier !== 1) errors.push('friction events must be tier:1');
  if (!SEVERITIES.includes(event.severity)) errors.push(`severity must be one of ${SEVERITIES.join('|')}`);
  return { ok: errors.length === 0, errors };
}

// G4: redact at the instrumentation site, never scrub at storage.
function emitFriction(patternId, fields = {}, opts = {}) {
  const file = opts.file || INCIDENTS_FILE;
  // redactEvent returns { event, hits }; we persist the redacted event only.
  const { event } = redactEvent(buildFrictionEvent(patternId, fields, opts.now));
  const { ok, errors } = isValidFriction(event);
  if (!ok) throw new Error(`invalid friction event: ${errors.join('; ')}`);
  emitV3(event, file);
  return event;
}

function loadFrictionCatalog(file = CATALOG_FILE) {
  try { return JSON.parse(fs.readFileSync(file, 'utf8')); } catch { return { patterns: [] }; }
}

function isKnownPattern(patternId, catalog = loadFrictionCatalog()) {
  return (catalog.patterns || []).some((entry) => entry.pattern_id === patternId);
}

module.exports = {
  FRICTION_EVENT, SEVERITIES, INCIDENTS_FILE, CATALOG_FILE,
  buildFrictionEvent, isValidFriction, emitFriction, loadFrictionCatalog, isKnownPattern,
};

File: hooks/scripts/friction_event.py
"""Per-team friction-event emitter (#3165, Epic #3008/#3095).

Python sibling of scripts/global/friction-event.js. emit_friction writes a REDACTED
schema-v3 event tagged ``tier: 1`` + ``pattern_id`` + ``severity`` to
``~/.megingjord/incidents.jsonl`` so the existing anneal-tier2-autofile recurrence
model promotes recurring frictions to Tier-2. Redaction mirrors the subset in
baton_event_emitter.py (G4: redact at the instrumentation site, not at storage).
"""
from __future__ import annotations

import json
import re
import time
from pathlib import Path
from typing import Any, Optional

SCHEMA_VERSION = 3
FRICTION_EVENT = "governance.friction"
SEVERITIES = ("low", "medium", "high", "critical")
DEFAULT_SEVERITY = "low"
INCIDENTS_LOG_PATH = Path.home() / ".megingjord" / "incidents.jsonl"

_REDACTION_PATTERNS = [
    (re.compile(r"sk-ant-[A-Za-z0-9_-]{20,}"), "<ANTHROPIC_KEY_REDACTED>"),
    (re.compile(r"sk-(?!ant-)[A-Za-z0-9_-]{20,}"), "<OPENAI_KEY_REDACTED>"),
    (re.compile(r"gh[pousr]_[A-Za-z0-9]{20,}"), "<GITHUB_PAT_REDACTED>"),
    (re.compile(r"Bearer [A-Za-z0-9_.-]{20,}"), "Bearer <REDACTED>"),
    (re.compile(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b"), "<EMAIL_REDACTED>"),
]


def _redact(value: Any) -> Any:
    """Recursively redact secret-bearing strings; structure is preserved."""
    if isinstance(value, str):
        out = value
        for pattern, replacement in _REDACTION_PATTERNS:
            out = pattern.sub(replacement, out)
        return out
    if isinstance(value, dict):
        return {k: _redact(v) for k, v in value.items()}
    if isinstance(value, list):
        return [_redact(v) for v in value]
    return value


def build_friction_event(pattern_id: str, fields: Optional[dict] = None,
                         now: Optional[str] = None) -> dict:
    """Build the schema-v3 friction event (matches the JS buildFrictionEvent shape)."""
    fields = fields or {}
    severity = fields.get("severity")
    if severity not in SEVERITIES:
        severity = DEFAULT_SEVERITY
    event = {
        "version": SCHEMA_VERSION,
        "ts": now or time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "service": "friction",
        "env": fields.get("env", "local"),
        "event": FRICTION_EVENT,
        "tier": 1,
        "pattern_id": pattern_id,
        "team": fields.get("team", "unknown"),
        "runtime": fields.get("runtime", "unknown"),
        "role": fields.get("role"),
        "surface": fields.get("surface"),
        "severity": severity,
        "workaround": fields.get("workaround"),
        "trigger_role": fields.get("role") or "system",
        "trigger_type": "friction",
    }
    if fields.get("cost") is not None:
        event["cost"] = fields["cost"]
    if fields.get("detail"):
        event["detail"] = fields["detail"]
    return event


def is_valid_friction(event: dict) -> bool:
    """True when the event is a well-formed friction event."""
    return bool(
        isinstance(event, dict)
        and event.get("event") == FRICTION_EVENT
        and event.get("pattern_id")
        and event.get("tier") == 1
        and event.get("severity") in SEVERITIES
    )


def emit_friction(pattern_id: str, fields: Optional[dict] = None,
                  path: Optional[Path] = None) -> Optional[dict]:
    """Append one redacted friction event to incidents.jsonl. Never raises."""
    try:
        target = Path(path) if path else INCIDENTS_LOG_PATH
        event = _redact(build_friction_event(pattern_id, fields))
        if not is_valid_friction(event):
            return None
        target.parent.mkdir(parents=True, exist_ok=True)
        with target.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(event) + "\n")
        return event
    except Exception:
        return None  # observability emission must never break a caller
